File: plugins/plugin-cron/src/actions/delete-cron.ts
/**
 * @module delete-cron
 * @description Action to delete a cron job
 */

import type { Action, IAgentRuntime, Memory, State, HandlerCallback } from '@elizaos/core';
import { CronService } from '../services/cron-service.js';
import { CRON_SERVICE_TYPE, CronActions } from '../constants.js';

/**
 * Extracts job identifier (ID or name) from text
 */
function extractJobIdentifier(text: string): { id?: string; name?: string } {
  // Try to find UUID
  const idMatch = /([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})/i.exec(text);
  if (idMatch) {
    return { id: idMatch[1] };
  }

  // Try to find quoted name
  const quotedMatch = /["']([^"']+)["']/i.exec(text);
  if (quotedMatch) {
    return { name: quotedMatch[1] };
  }

  // Try to find "job called X" or "cron named X"
  const namedMatch = /(?:job|cron)\s+(?:called|named)\s+(\S+)/i.exec(text);
  if (namedMatch) {
    return { name: namedMatch[1] };
  }

  return {};
}

export const deleteCronAction: Action = {
  name: CronActions.DELETE_CRON,
  similes: [
    'REMOVE_CRON',
    'CANCEL_CRON',
    'STOP_CRON',
    'DELETE_SCHEDULED_JOB',
    'REMOVE_SCHEDULED_JOB',
  ],
  description: 'Deletes a cron job by ID or name, removing it from the schedule permanently.',

  validate: async (_runtime: IAgentRuntime, message: Memory): Promise<boolean> => {
    const text = message.content?.text?.toLowerCase() ?? '';

    const hasDeleteKeyword =
      text.includes('delete') ||
      text.includes('remove') ||
      text.includes('cancel') ||
      (text.includes('stop') && !text.includes('stop running'));

    const hasCronKeyword =
      text.includes('cron') || text.includes('job') || text.includes('schedule');

    return hasDeleteKeyword && hasCronKeyword;
  },

  handler: async (
    runtime: IAgentRuntime,
    message: Memory,
    _state?: State,
    options?: Record<string, unknown>,
    callback?: HandlerCallback
  ) => {
    const cronService = runtime.getService<CronService>(CRON_SERVICE_TYPE);
    if (!cronService) {
      await callback?.({
        text: 'Cron service is not available. Please ensure the plugin is loaded.',
      });
      return { success: false, error: 'Cron service not available' };
    }

    const text = message.content?.text ?? '';

    // Get job ID from options or extract from text
    let jobId: string | undefined = options?.jobId as string | undefined;
    let jobName: string | undefined;

    if (!jobId) {
      const identifier = extractJobIdentifier(text);

      if (identifier.id) {
        jobId = identifier.id;
      } else if (identifier.name) {
        // Find job by name
        const jobs = await cronService.listJobs({ includeDisabled: true });
        const job = jobs.find(
          (j) => j.name.toLowerCase() === identifier.name!.toLowerCase()
        );

        if (!job) {
          await callback?.({
            text: `No cron job found with name: ${identifier.name}`,
          });
          return { success: false, error: 'Job not found' };
        }

        jobId = job.id;
        jobName = job.name;
      }
    }

    if (!jobId) {
      await callback?.({
        text:
          'Please specify which cron job to delete. You can use the job ID or name.\n' +
          'Example: "delete cron job abc-123" or "remove cron called daily-check"',
      });
      return { success: false, error: 'No job identifier provided' };
    }

    // Get job details before deletion for confirmation message
    if (!jobName) {
      const job = await cronService.getJob(jobId);
      if (job) {
        jobName = job.name;
      }
    }

    const deleted = await cronService.deleteJob(jobId);

    if (!deleted) {
      await callback?.({
        text: `No cron job found with ID: ${jobId}`,
      });
      return { success: false, error: 'Job not found' };
    }

    await callback?.({
      text: `Deleted cron job "${jobName || 'unknown'}" (${jobId}).\nThe job has been permanently removed and will no longer run.`,
    });

    return {
      success: true,
      data: {
        jobId,
        jobName,
        deleted: true,
      },
    };
  },

  examples: [
    [
      {
        name: '{{user1}}',
        content: { text: 'Delete the cron job called daily-check' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Deleted cron job "daily-check" (abc-123).\nThe job has been permanently removed and will no longer run.',
        },
      },
    ],
    [
      {
        name: '{{user1}}',
        content: { text: 'Remove cron abc-123-def-456' },
      },
      {
        name: '{{agentName}}',
        content: {
          text: 'Deleted cron job "hourly-status" (abc-123-def-456).\nThe job has been permanently removed and will no longer run.',
        },
      },
    ],
  ],
};

File: plugins/plugin-cron/rust/src/actions/delete_cron.rs
use async_trait::async_trait;
use serde_json::Value;

use crate::{Action, ActionExample, ActionResult, CronService};

pub struct DeleteCronAction;

#[async_trait]
impl Action for DeleteCronAction {
    fn name(&self) -> &str {
        "DELETE_CRON"
    }

    fn similes(&self) -> Vec<&str> {
        vec![
            "REMOVE_CRON",
            "CANCEL_CRON",
            "STOP_CRON",
            "DELETE_SCHEDULED_JOB",
            "REMOVE_SCHEDULED_JOB",
        ]
    }

    fn description(&self) -> &str {
        "Deletes a cron job by ID or name, removing it from the schedule permanently."
    }

    async fn validate(&self, message: &Value, _state: &Value) -> bool {
        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("")
            .to_lowercase();

        let has_delete = text.contains("delete")
            || text.contains("remove")
            || text.contains("cancel")
            || (text.contains("stop") && !text.contains("stop running"));

        let has_cron = text.contains("cron") || text.contains("job") || text.contains("schedule");

        has_delete && has_cron
    }

    async fn handler(
        &self,
        message: &Value,
        _state: &Value,
        service: Option<&mut CronService>,
    ) -> ActionResult {
        let service = match service {
            Some(s) => s,
            None => {
                return ActionResult {
                    success: false,
                    text: "Cron service is not available.".to_string(),
                    data: None,
                    error: Some("missing_service".to_string()),
                }
            }
        };

        let text = message
            .get("content")
            .and_then(|c| c.get("text"))
            .and_then(|t| t.as_str())
            .unwrap_or("");

        let job_id = extract_job_id_for_delete(text, service);

        let job_id = match job_id {
            Some(id) => id,
            None => {
                return ActionResult {
                    success: false,
                    text: "Please specify which cron job to delete (by ID or name).".to_string(),
                    data: None,
                    error: Some("No job identifier".to_string()),
                }
            }
        };

        // Get name before deleting
        let job_name = service
            .get_job(&job_id)
            .map(|j| j.name.clone())
            .unwrap_or_else(|| "unknown".to_string());

        match service.delete_job(&job_id) {
            Ok(true) => ActionResult {
                success: true,
                text: format!(
                    "Deleted cron job \"{}\" ({}).\nThe job has been permanently removed.",
                    job_name, job_id
                ),
                data: Some(serde_json::json!({
                    "jobId": job_id,
                    "jobName": job_name,
                    "deleted": true
                })),
                error: None,
            },
            Ok(false) => ActionResult {
                success: false,
                text: format!("No cron job found with ID: {}", job_id),
                data: None,
                error: Some("Job not found".to_string()),
            },
            Err(e) => ActionResult {
                success: false,
                text: format!("Failed to delete job: {}", e),
                data: None,
                error: Some(e),
            },
        }
    }

    fn examples(&self) -> Vec<ActionExample> {
        vec![
            ActionExample {
                user_message: "Delete the cron job called daily-check".to_string(),
                agent_response: "Deleted cron job \"daily-check\" (abc-123).\nThe job has been permanently removed.".to_string(),
            },
            ActionExample {
                user_message: "Remove cron abc-123-def-456".to_string(),
                agent_response: "Deleted cron job \"hourly-status\" (abc-123-def-456).\nThe job has been permanently removed.".to_string(),
            },
        ]
    }
}

fn extract_job_id_for_delete(text: &str, service: &CronService) -> Option<String> {
    // UUID
    let uuid_re = regex::Regex::new(
        r"(?i)([a-f0-9]{8}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{4}-[a-f0-9]{12})",
    )
    .ok()?;
    if let Some(caps) = uuid_re.captures(text) {
        return Some(caps.get(1)?.as_str().to_string());
    }

    // Quoted name
    let quoted_re = regex::Regex::new(r#"["']([^"']+)["']"#).ok()?;
    if let Some(caps) = quoted_re.captures(text) {
        let name = caps.get(1)?.as_str();
        if let Some(job) = service.find_job_by_name(name) {
            return Some(job.id.clone());
        }
    }

    // "called/named X"
    let named_re = regex::Regex::new(r"(?i)(?:called|named)\s+(\S+)").ok()?;
    if let Some(caps) = named_re.captures(text) {
        let name = caps.get(1)?.as_str();
        if let Some(job) = service.find_job_by_name(name) {
            return Some(job.id.clone());
        }
    }

    None
}
